File: src/test/java/org/effective_mobile/task_management_system/resource/AuthenticationResourceTest.java
package org.effective_mobile.task_management_system.resource;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.effective_mobile.task_management_system.component.UsernameProvider;
import org.effective_mobile.task_management_system.confings.IntegrationTest;
import org.effective_mobile.task_management_system.database.entity.User;
import org.effective_mobile.task_management_system.resource.json.UserCreationResponsePojo;
import org.effective_mobile.task_management_system.resource.json.auth.SigninRequestPojo;
import org.effective_mobile.task_management_system.resource.json.auth.SignupRequestPojo;
import org.effective_mobile.task_management_system.security.AuthTokenComponent;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.platform.commons.util.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.jdbc.EmbeddedDatabaseConnection;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.http.MediaType;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;

import java.util.Optional;

import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@TestPropertySource(properties = {
    "spring.jpa.hibernate.ddl-auto=create-drop"
})
@AutoConfigureTestDatabase(connection = EmbeddedDatabaseConnection.H2)
public class AuthenticationResourceTest extends IntegrationTest {

    @Autowired private ObjectMapper objectMapper;
    @Autowired private AuthTokenComponent authTokenComponent;
    @Autowired private UsernameProvider usernameProvider;

    /**
     * Test for {@link AuthenticationResource#signin} и {@link AuthenticationResource#signup}.
     */
    @Test
    public void signInSignUpTest() throws Exception {
        testSignUp();
        testSignIn();
    }

    private void testSignUp() throws Exception {
        SignupRequestPojo signupRequestPojo = new SignupRequestPojo(email, username, password);
        MvcResult mvcResult = post(Api.SIGN_UP, signupRequestPojo).andReturn();

        UserCreationResponsePojo userCreationResponsePojo = objectMapper.readValue(mvcResult.getResponse().getContentAsString(), UserCreationResponsePojo.class);

        User user = userRepository.findOrThrow(User.class, userCreationResponsePojo.getId());

        Assertions.assertTrue(passwordEncoder.matches(password, user.getPassword()));
        Assertions.assertEquals(user.getEmail(), email);
        Assertions.assertEquals(user.getUsername(), username);
    }

    private void testSignIn() throws Exception {
        SigninRequestPojo signinRequestPojo = new SigninRequestPojo(email, password);
        MvcResult resultActions =
            post(Api.SIGN_IN, signinRequestPojo)
                .andExpect(content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
                .andReturn();

        final JsonNode node = objectMapper.readTree(resultActions.getResponse().getContentAsString());

        JsonNode authTokenJsonNode = node.get(authTokenNameLazy.get());
        Assertions.assertNotNull(authTokenJsonNode);
        String token = authTokenJsonNode.asText();
        Assertions.assertTrue(StringUtils.isNotBlank(token));

        Optional<User> optionalUser = userRepository.findByEmail(email);
        Assertions.assertTrue(optionalUser.isPresent());

        User user = optionalUser.get();
        String subject = authTokenComponent.validateTokenAndGetUsername(token);
        Assertions.assertEquals(usernameProvider.getUsername(user), subject);
        Assertions.assertTrue(passwordEncoder.matches(password, user.getPassword()));
    }

    private ResultActions post(String signin, Object body) throws Exception {
        return mvc
            .perform(MockMvcRequestBuilders
                .post(signin)
                .contentType(MediaType.APPLICATION_JSON)
                .content(asString(body))
            )
            .andExpect(status().isOk());
    }
}

File: src/test/kotlin/org/effective_mobile/task_management_system/resource/AuthenticationResourceTest.kt
package org.effective_mobile.task_management_system.resource

import org.effective_mobile.task_management_system.AssertionsUtils
import org.effective_mobile.task_management_system.RandomUsers
import org.effective_mobile.task_management_system.database.entity.User
import org.effective_mobile.task_management_system.database.repository.PrivilegeRepository
import org.effective_mobile.task_management_system.database.repository.RoleRepository
import org.effective_mobile.task_management_system.database.repository.TaskRepository
import org.effective_mobile.task_management_system.database.repository.UserRepository
import org.effective_mobile.task_management_system.pojo.HasUserInfo
import org.effective_mobile.task_management_system.resource.json.assignment.SignupResponsePojo
import org.effective_mobile.task_management_system.resource.json.auth.SigninRequestPojo
import org.effective_mobile.task_management_system.resource.json.auth.SigninResponsePojo
import org.effective_mobile.task_management_system.resource.json.auth.SignupRequestPojo
import org.effective_mobile.task_management_system.security.authentication.AuthenticationTokenComponent
import org.effective_mobile.task_management_system.utils.Api
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.platform.commons.util.StringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpMethod.POST
import org.springframework.security.crypto.password.PasswordEncoder

class AuthenticationResourceTest @Autowired constructor(
    userRepository: UserRepository,
    taskRepository: TaskRepository,
    privilegeRepository: PrivilegeRepository,
    roleRepository: RoleRepository
): UserAndTaskIntegrationBase(
    userRepository,
    taskRepository,
    privilegeRepository,
    roleRepository
) {
    @Autowired
    private lateinit var authenticationTokenComponent: AuthenticationTokenComponent

    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    /**
     * Test for [AuthenticationResource.signup].
     */
    @Test
    @Throws(Exception::class)
    fun testSignUp() {
        val user = RandomUsers.user()

        user {
            unauthenticated()
            send(mvc) {
                method = POST
                url = Api.SIGN_UP
                body = SignupRequestPojo(getEmail(), getUsername(), getPassword())
            } response {
                val signupResponsePojo = getBody<SignupResponsePojo>()
                userRepository.findOrThrow(User::class.java, signupResponsePojo.id).let { foundUser ->
                    Assertions.assertTrue(passwordEncoder.matches(getPassword(), foundUser.getPassword()))

                    AssertionsUtils.assertEquals(user, foundUser, HasUserInfo::getEmail)
                    AssertionsUtils.assertEquals(user, foundUser, HasUserInfo::getUsername)
                }
            }
        }
    }

    /**
     * Test for [AuthenticationResource.signin].
     */
    @Test
    @Throws(Exception::class)
    fun testSignIn() {
        val notEncodedPassword = RandomUsers.safePassword()
        val user = saveAndFlush(RandomUsers.user { passwordEncoder.encode(notEncodedPassword) })

        user {
            send(mvc) {
                method = POST
                url = Api.SIGN_IN
                body = SigninRequestPojo(getEmail(), notEncodedPassword)
            } response {
                val authToken = getBody<SigninResponsePojo>().authToken
                Assertions.assertNotNull(authToken)
                Assertions.assertTrue(StringUtils.isNotBlank(authToken))

                userRepository.findByEmail(getEmail()).apply {
                    val foundUser = get()
                    Assertions.assertNotNull(foundUser)
                    val username = authenticationTokenComponent.validateTokenAndGetUsername(authToken)
                    Assertions.assertEquals(usernameProvider.getUsername(foundUser), username)
                    Assertions.assertTrue(passwordEncoder.matches(notEncodedPassword, foundUser.getPassword()))
                }
            }
        }
    }

    //TODO: write test for signin where invalid credentials are present
}
